pkg/infoschema/infoschema_v2.go

```diff
@@ -0,0 +1,431 @@
+		if !ok { // The first one found.
+			ok = true
+			itm = item
+		} else { // The latest one
+			if item.schemaVersion > itm.schemaVersion {
+				itm = item
+			}
+		}
+		return true
+	})
+	return itm, ok
+}
+
+func (is *infoschemaV2) TableByID(id int64) (val table.Table, ok bool) {
+	if isTableVirtual(id) {
+		return nil, false
+	}
+
+	// Get from the cache.
+	key := cacheKey{id, is.schemaVersion}
+	tbl, found := is.cache.Get(key)
+	if found && tbl != nil {
+		return tbl, true
+	}
+
+	eq := func(a, b *Item) bool { return a.tableID == b.tableID }
+	itm, ok := search(is.byID, is.schemaVersion, Item{tableID: id, dbID: math.MaxInt64}, eq)
+	if !ok {
+		return nil, false
+	}
+
+	// Maybe the table is evicted? need to reload.
+	ret, err := loadTableInfo(is.r, is.Data, id, itm.dbID, is.ts)
+	if err == nil {
+		is.cache.Set(key, ret)
+		return ret, true
+	}
+	return nil, false
+}
+
+func (is *infoschemaV2) TableByName(schema, tbl model.CIStr) (t table.Table, err error) {
+	if schema.L == "information_schema" || schema.L == "metrics_schema" || schema.L == "performance_schema" {
+		if tbNames, ok := is.specials[schema.L]; ok {
+			if t, ok = tbNames.tables[tbl.L]; ok {
+				return
+			}
+		}
+		return nil, ErrTableNotExists.GenWithStackByArgs(schema, tbl)
+	}
+
+	eq := func(a, b *Item) bool { return a.dbName == b.dbName && a.tableName == b.tableName }
+	itm, ok := search(is.name2id, is.schemaVersion, Item{dbName: schema.L, tableName: tbl.L, tableID: math.MaxInt64}, eq)
+	if !ok {
+		// TODO: in the future, this may happen and we need to check tikv to see whether table exists.
+		return nil, ErrTableNotExists.GenWithStackByArgs(schema, tbl)
+	}
+
+	// Get from the cache.
+	key := cacheKey{itm.tableID, is.schemaVersion}
+	res, found := is.cache.Get(key)
+	if found && res != nil {
+		return res, nil
+	}
+
+	// Maybe the table is evicted? need to reload.
+	ret, err := loadTableInfo(is.r, is.Data, itm.tableID, itm.dbID, is.ts)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	is.cache.Set(key, ret)
+	return ret, nil
+}
+
+func (is *infoschemaV2) SchemaByName(schema model.CIStr) (val *model.DBInfo, ok bool) {
+	var dbInfo model.DBInfo
+	dbInfo.Name = schema
+	is.schemaMap.Descend(schemaItem{dbInfo: &dbInfo, schemaVersion: math.MaxInt64}, func(item schemaItem) bool {
+		if item.Name() != schema.L {
+			ok = false
+			return false
+		}
+		if item.schemaVersion <= is.schemaVersion {
+			ok = true
+			val = item.dbInfo
+			return false
+		}
+		return true
+	})
+	return
+}
+
+func (is *infoschemaV2) AllSchemas() (schemas []*model.DBInfo) {
+	is.schemaMap.Scan(func(item schemaItem) bool {
+		// TODO: version?
+		schemas = append(schemas, item.dbInfo)
+		return true
+	})
+	return
+}
+
+func (is *infoschemaV2) SchemaMetaVersion() int64 {
+	return is.schemaVersion
+}
+
+func (is *infoschemaV2) SchemaExists(schema model.CIStr) bool {
+	var ok bool
+	// TODO: support different version
+	is.schemaMap.Scan(func(item schemaItem) bool {
```

Comment:
use `Descend`?

Reply:
This is a full memory `Scan`, it does not matter whether we `Descend` or `Ascend`

Reply:
seems `Descend` can only search items from [pivot, first], no need a full scan